runtime/src/accounts_db.rs

```diff
@@ -2040,6 +2240,218 @@ impl AccountsDB {
+                    assert_eq!(reclaimed_slot, removed_slot);
+                    assert_eq!(
+                        reclaimed_account_info.store_id,
+                        Self::cache_store_virtual_id()
+                    );
+                }
+            }
+        }
+
+        // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        for old_slot in self.accounts_cache.find_older_frozen_slots(MAX_CACHE_SLOTS) {
+            self.flush_slot_cache(old_slot);
+        }
+
+        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
+    }
+
+    fn flush_slot_cache(&self, slot: Slot) {
+        info!("flush_slot_cache slot: {}", slot);
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            let iter_items: Vec<_> = slot_cache.iter().collect();
+            let mut total_size = 0;
+            let (accounts, hashes): (Vec<(&Pubkey, &Account)>, Vec<Hash>) = iter_items
+                .iter()
+                .map(|iter_item| {
+                    let key = iter_item.key();
+                    let account = &iter_item.value().account;
+                    let hash = iter_item.value().hash;
+                    total_size += (account.data.len() + STORE_META_OVERHEAD) as u64;
+                    ((key, account), hash)
+                })
+                .unzip();
+            let aligned_total_size = self.page_align(total_size);
+            let flushed_store =
+                self.create_and_insert_store(slot, aligned_total_size, "flush_slot_cache");
+            self.store_accounts_custom(
+                slot,
+                &accounts,
+                &hashes,
+                Some(Box::new(move |_, _| flushed_store.clone())),
+                None,
+                false,
+            );
+            // If the above sizing function is correct, just one AppendVec is enough to hold
+            // all the data for the slot
+            assert_eq!(
+                self.storage
+                    .get_slot_stores(slot)
+                    .unwrap()
+                    .read()
+                    .unwrap()
+                    .len(),
+                1
+            );
+        };
+
+        self.accounts_cache.remove_slot(slot);
+    }
+
+    // A specially reserved storage id just for entries in the cache, so that
+    // operations that take a storage entry can maintain a common interface
+    // when interacting with cached accounts. This id is "virtual" in that it
+    // doesn't actually refer to an actual storage entry.
+    fn cache_store_virtual_id() -> usize {
+        AppendVecId::MAX
+    }
+
+    fn cache_accounts(
+        &self,
+        slot: Slot,
+        hashes: &[Hash],
+        accounts_and_meta_to_store: &[(StoredMeta, &Account)],
+    ) -> Vec<AccountInfo> {
+        assert_eq!(hashes.len(), accounts_and_meta_to_store.len());
+        accounts_and_meta_to_store
+            .iter()
```

Comment:
how about using `iter_par`?

Reply:
iirc, or dashmap isn't that great for concurrent writes?

Reply:
@ryoqun, I think actually Dashmap should be the best out of the concurrent HashMaps for concurrent writes. I didn't use `iter_par` here because there should already be multiple threads committing/storing their own set of transactions.

Reply:
ok, thanks for explaining. :) yeah, that makes sense.